### DISCUSSION TITLE
How do you create custom component in v7?

### DISCUSSION THREAD

**Discussion by Toshinaki** [mantinedev/mantine#5062] [Q&A]
In v6 we could create [custom component](https://v6.mantine.dev/guides/custom-components/). With codes below, we can create component that accepts `classNames`, `styles`, .etc:
```
type MyComponentStylesNames = Selectors<typeof useStyles>;

interface MyComponentProps extends DefaultProps<MyComponentStylesNames, MyComponentStylesParams> {
  radius?: MantineNumberSize;
}
```

How to achieve this in v7?

For example for this component:
```
import 'styles.module.css';

...

<div className={classes.root}>
  <div className={classes.wrapper}>
  </div>
</div>
```
do I have to define the component props manually and merge with clsx?
```
interface Props {
  classNames?: {
    root?: string
    wrapper?: string
  }
}

...
<div className={clsx(classes.root, classNames.root)}>
  <div className={clsx(classes.wrapper, classNames.wrapper)}>
  </div>
</div>
```

**Comment by rtivital** [comment]
You can explore Mantine components source code to learn how to build custom components. The guide is planned to be added, but not soon.

**Reply by Toshinaki** [reply]
[StylesApiProps](https://github.com/mantinedev/mantine/blob/0392439c21334f39cd13c8051c3cab600baf9ac2/src/mantine-core/src/core/styles-api/styles-api.types.ts#L46)? Not sure how to implement with this.
Could you please point to a simple and straightforward enough component to learn from? Or a simple example?

**Reply by rtivital** [reply]
You can use AspectRatio as a simple example

**Reply by Toshinaki** [reply]
@rtivital 
Thanks! I created a component successfully using AspectRatio as example.
If I want to wrap the component with `memo`, should I do
```
const MyComp = memo(factory<MyCompFactory>((_props, ref) => {
})
```
or in a reverse order? (`factory(memo(...))`)

**Reply by Toshinaki** [reply]
I've got another question.  Since it's still about creating custom component, I'll just ask here.

I use `...getStyles('className')` to apply `className` and `style` to the component, as the example does.
If I have a list of links:
```
<Link className={clsx(classes.link, classes.home)}>Home</Link>
<Link className={classes.link}>blog</Link>
<Link className={classes.link}>about</Link>
```
with `getStyles`, can be write as:
```
<Link className={clsx(classes.link, classes.home)}>Home</Link>
<Link ...getStyles('link')>blog</Link>
<Link ...getStyles('link')>about</Link>
```
Is there any way to use `getStyles` for `clsx(classes.link, classes.home)`?

**Reply by rtivital** [reply]
`getStyles` requires one className per element, modifiers are supposed to be added with data attributes, see https://mantine.dev/styles/data-attributes/

**Reply by Toshinaki** [reply]
@rtivital 
I've tried data attributes, and `Box`'s `mod` attribute is sometimes useful.
But to use it on other components, I have to do `<Box component={...} mod={{...}}>`
Is it possible for you to modify the implementation of `useStyles` so that `getStyles` can take an array of class names?
Could I post a feature request?

**Reply by rtivital** [reply]
If you are building custom components and want to propose improvements to API, you are welcome to create a separate discussion for that with more details

**Reply by tombohub** [reply]
@Toshinaki Can you post or link an example how did you make your component?
